autogpts/autogpt/autogpt/commands/execute_code.py

```diff
@@ -197,71 +198,81 @@ def execute_python_file(self, filename: str, args: list[str] | str = []) -> str:
-                            logger.info(f"{status}: {progress}")
-                        elif status:
-                            logger.info(status)
-
-                logger.debug(f"Creating new {image_name} container...")
-                container: DockerContainer = client.containers.run(
-                    image_name,
-                    ["sleep", "60"],  # Max 60 seconds to prevent permanent hangs
-                    volumes={
-                        str(self.workspace.root): {
-                            "bind": "/workspace",
-                            "mode": "rw",
-                        }
-                    },
-                    working_dir="/workspace",
+                    container: DockerContainer = client.containers.get(
+                        container_name
+                    )  # type: ignore
+                except NotFound:
+                    try:
+                        client.images.get(image_name)
+                        logger.debug(f"Image '{image_name}' found locally")
+                    except ImageNotFound:
+                        logger.info(
+                            f"Image '{image_name}' not found locally,"
+                            " pulling from Docker Hub..."
+                        )
+                        # Use the low-level API to stream the pull response
+                        low_level_client = docker.APIClient()
+                        for line in low_level_client.pull(
+                            image_name, stream=True, decode=True
+                        ):
+                            # Print the status and progress, if available
+                            status = line.get("status")
+                            progress = line.get("progress")
+                            if status and progress:
+                                logger.info(f"{status}: {progress}")
+                            elif status:
+                                logger.info(status)
+
+                    logger.debug(f"Creating new {image_name} container...")
+                    container: DockerContainer = client.containers.run(
+                        image_name,
+                        ["sleep", "60"],  # Max 60 seconds to prevent permanent hangs
+                        volumes={
+                            str(Path(local_path).absolute()): {
+                                "bind": "/workspace",
+                                "mode": "rw",
+                            }
+                        },
+                        working_dir="/workspace",
+                        stderr=True,
+                        stdout=True,
+                        detach=True,
+                        name=container_name,
+                    )  # type: ignore
+                    container_is_fresh = True
+
+                if not container.status == "running":
+                    container.start()
+                elif not container_is_fresh:
+                    container.restart()
```

Comment:
Maybe abstract this into a helper method

Reply:
done